### PR TITLE
lightdm: allow cursor theme customisation. [WIP]

### PR DESCRIPTION
**Pull request by pbogdan**
###### Motivation for this change

Allow customisation of cursor theme (both theme and cursor size as well) in lightdm's gtk greeter.
With this change one can configure both the greeter and the desktop environment to have a consistent cursor theme. Also comes handy on a hidpi display where lightdm by default was using tiny cursor size.

This is both work in progress as well as a request for feedback to see whether there is interest in those changes.

In terms of things that would still need to be done: I would like to make the new settings optional so they don't affect existing setups unless users explicitly opt in.

Also to note - outside of changes to the nixos module itself this PR includes [a patch to lightdm's sources](https://github.com/NixOS/nixpkgs/compare/master...pbogdan:lightdm?expand=1#diff-936b708c2703c3e4cee18d748a2e167f)

Perhaps there is a less invasive way to accomplish what this PR is trying to do in which case I would like to hear about it 😄 

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/applications/display-managers/lightdm-gtk-greeter/default.nix [hunk 1]
```diff
@@ -24,6 +23,7 @@ stdenv.mkDerivation rec {
       url = "https://588764.bugs.gentoo.org/attachment.cgi?id=442616";
       sha256 = "0r383kjkvq9yanjc1lk878xc5g8993pjgxylqhhjb5rkpi1mbfsv";
     })
+    ./cursors.patch
```
Comment (Mic92) [inline pkgs/applications/display-managers/lightdm-gtk-greeter/default.nix]: Can you also submit this patch to upstream?

Comment (pbogdan) [inline pkgs/applications/display-managers/lightdm-gtk-greeter/default.nix]: Thanks, submitted upstream https://code.launchpad.net/~ppbogdan/lightdm-gtk-greeter/cursor-theme/+merge/324378

**Comment by pbogdan** [comment]
I'm yet to receive any feedback from upstream, while I'm not in a rush to get this merged (if there is interest) would the patch be suitable for inclusion in nixpkgs if it's not accepted by lightdm project?

**Comment by Balletie** [comment]
I've been thinking about this change myself too, however I never got around to it. So it's very welcome.

Maybe it's also possible by setting the system-wide default cursor? A quick google shows that the location for this is `/usr/share/icons/default/index.theme`, so perhaps for NixOS it's `/run/current-system/sw/share/icons/default/index/theme`.

**Comment by pbogdan** [comment]
There is an issue about setting up a global cursor theme https://github.com/NixOS/nixpkgs/issues/22652 however it doesn't sound like anyone's found a solution yet ☹️ (which is what partially motivated this patch though admittedly I haven't looked too much into setting up a global `default/index.theme`).

**Comment by pbogdan** [comment]
Playing with this some more using `default/index.theme` could work for setting the theme:

1. Add the following the `environment.systemPackages`:

    ```
    (pkgs.callPackage ({ stdenv }: stdenv.mkDerivation {
      name = "global-cursor-theme";
      unpackPhase = "true";
      outputs = [ "out" ];
      installPhase = ''
        mkdir -p $out/share/icons/default
        cat << EOF > $out/share/icons/default/index.theme
        [Icon Theme]
        Name=Default
        Comment=Default Cursor Theme
        Inherits=Oxygen_Blue
        EOF
      '';
    }) {})
    ```
2. Make sure the cursor theme that's being `Inherits`'ed from is installed system wide.
3. Tweak `XCURSOR_PATH` in  ` nixos/modules/services/x11/display-managers/lightdm-greeters/gtk.nix` as per this PR sans `lib.optionalString` and set it to `/run/current-system/sw/share/icons/`

As I understand it would have the benefit of setting the theme for applications too though would still leave the issue of setting the size.

**Comment by pbogdan** [comment]
Further experimenting reveals `XCURSOR_SIZE` env variable that would appear to be respected by lightdm 😆

**Comment by Balletie** [comment]
Thanks for your experiments! :)

**Pull request closed without merging**